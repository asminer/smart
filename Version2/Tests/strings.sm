
// $Id$

/*
    Test of string operators and such
*/


/* String addition: */

string first := "This should be con";
string second := "nected";

print("string 1: ", first, "\n");
print("string 2: ", second, "\n");
print("sum     : ", first+second, "\n");

/* Empty string test */

string empty := "";

print("Empty string: --->", empty, "<---\n");

print("a"+empty, "\n");
print(empty+"a", "\n");
print("a"+empty+"a", "\n");
print("a"+empty+"a\n");

/* String function test */

string nrpt(string x, int n) := cond(n<1, "", x+nrpt(x, n-1));
                                       
for(int i in {-2 .. 25}) {
  print("--->", nrpt("#", i), "<---\n");
}

string repeat(string x, int n) := cond(n<1, "",
					cond( mod(n,2)==0, 
					      repeat(x+x, div(n,2)),
					      x+repeat(x, n-1)
                                            )
			              );


for(int i in {25..-2..-1}) {
  print("--->", repeat("x", i), "<---\n");
}

/* substr tests */

string full := "This is a test string";

print(substr(full, 0, 0), "\n");
print(substr(full, -5, 0), "\n");
print(substr(full, -infinity, 0), "\n");
print(substr(full, 0, 3), "\n");
print(substr(full, -5, 3), "\n");
print(substr(full, -infinity, 3), "\n");
print(substr(full, 2, 7), "\n");
print(substr(full, 2, 70), "\n");
print(substr(full, 2, infinity), "\n");
print(substr(full, -infinity, infinity), "\n");
print("Several empty substrings:\n");
print(substr(full, 2, 1), "\n");
print(substr(full, 70, 1), "\n");
print(substr(full, 70, 75), "\n");
print(substr(full, infinity, 1), "\n");
print(substr(full, 2, -5), "\n");
print(substr(full, -19, -5), "\n");
print(substr(full, 2, -infinity), "\n");
print(substr(full, -2, -infinity), "\n");
print(substr(full, infinity, -infinity), "\n");
print(substr(full, infinity, infinity), "\n");
print(substr(full, -infinity, -infinity), "\n");
print("Did you see them?\n");
