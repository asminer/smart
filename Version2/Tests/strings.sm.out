string 1: This should be con
string 2: nected
sum     : This should be connected
Empty string: ---><---
a
a
aa
aa
---><---
---><---
---><---
--->#<---
--->##<---
--->###<---
--->####<---
--->#####<---
--->######<---
--->#######<---
--->########<---
--->#########<---
--->##########<---
--->###########<---
--->############<---
--->#############<---
--->##############<---
--->###############<---
--->################<---
--->#################<---
--->##################<---
--->###################<---
--->####################<---
--->#####################<---
--->######################<---
--->#######################<---
--->########################<---
--->#########################<---
--->xxxxxxxxxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxxx<---
--->xxxxxxxxxxxxx<---
--->xxxxxxxxxxxx<---
--->xxxxxxxxxxx<---
--->xxxxxxxxxx<---
--->xxxxxxxxx<---
--->xxxxxxxx<---
--->xxxxxxx<---
--->xxxxxx<---
--->xxxxx<---
--->xxxx<---
--->xxx<---
--->xx<---
--->x<---
---><---
---><---
---><---
T
T
T
This
This
This
is is 
is is a test string
is is a test string
This is a test string
Several empty substrings:











Did you see them?
